src/py.rs

```diff
@@ -0,0 +1,305 @@
+use std::cmp::Ordering;
+use std::collections::btree_map::IntoIter;
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+use std::iter;
+
+use pyo3::basic::CompareOp;
+use pyo3::class::iter::{IterNextOutput, PyIterProtocol};
+use pyo3::conversion::{ToBorrowedObject, ToPyObject};
+use pyo3::exceptions::PyKeyError;
+use pyo3::types::PyDict;
+use pyo3::{prelude::*, PyMappingProtocol, PyNativeType, PyObjectProtocol};
+
+type Hash = isize;
+
+#[pyclass]
+#[derive(Debug, Clone, Default)]
+pub struct Dict {
+    map: BTreeMap<Hash, BTreeMap<Hash, (PyObject, PyObject)>>,
+}
+
+impl TryFrom<&PyDict> for Dict {
+    type Error = PyErr;
+    fn try_from(dict: &PyDict) -> PyResult<Self> {
+        let mut map = Dict::new();
+        for (k, v) in dict {
+            map.set_item(dict.py(), k, v)?;
+        }
+        Ok(map)
+    }
+}
+
+fn as_dict_obj(py: Python, obj: PyObject) -> PyResult<PyObject> {
+    let obj = obj.into_ref(py);
+    let obj = if obj.is_instance::<PyDict>()? {
+        Dict::try_from(obj.downcast::<PyDict>()?)?.into_py(py)
+    } else {
+        obj.to_object(py)
+    };
+    Ok(obj)
+}
+
+impl Dict {
+    fn is_eq(py: Python, a: &PyObject, b: &PyObject) -> bool {
+        a.as_ref(py)
+            .compare(b.as_ref(py))
+            .map(|ord| Ordering::Equal == ord)
+            .unwrap_or(false)
+    }
+
+    pub fn set_item<K: ToPyObject, V: ToPyObject>(
```

Comment:
```suggestion
    pub fn set<K: ToPyObject, V: ToPyObject>(
```

Reply:
Or are these methods required to have these names?

Reply:
No they are not, just a python convention.